src/query/service/src/sql/planner/binder/select.rs

```diff
@@ -335,7 +335,11 @@ impl<'a> Binder {
         right_expr: SExpr,
         distinct: bool,
     ) -> Result<SExpr> {
-        let union_plan = UnionAll {};
+        let mut column2type = HashMap::with_capacity(bind_context.columns.len());
+        for column_binding in bind_context.columns.iter() {
+            column2type.insert(column_binding.index, column_binding.data_type.clone());
+        }
+        let union_plan = UnionAll { column2type };
```

Comment:
It's better to add new column entries to metadata instead of modifying the data type here.

Reply:
Here I just store <column_idx, column_data_type>, then used in building physical plan, not modifying the data type.

Reply:
Then the data type can be inferred from input schema?

Reply:
Then we'll generate new schema as union's final schema from union's left child's output schema, the new schema's fields' data type will use data type of `column2type`.

Reply:
Make sense, will try using `EvalScalar` to cast type.